pkg/compose/build_buildkit.go

```diff
@@ -17,76 +17,132 @@
+	Targets []string `json:"targets"`
+}
+
+type bakeTarget struct {
+	Context    string            `json:"context,omitempty"`
+	Dockerfile string            `json:"dockerfile,omitempty"`
+	Args       map[string]string `json:"args,omitempty"`
+	Labels     map[string]string `json:"labels,omitempty"`
+	Tags       []string          `json:"tags,omitempty"`
+	CacheFrom  []string          `json:"cache-from,omitempty"`
+	CacheTo    []string          `json:"cache-to,omitempty"`
+	Secrets    []string          `json:"secret,omitempty"`
+	SSH        []string          `json:"ssh,omitempty"`
+	Platforms  []string          `json:"platforms,omitempty"`
+	Target     string            `json:"target,omitempty"`
+	Pull       bool              `json:"pull,omitempty"`
+	NoCache    bool              `json:"no-cache,omitempty"`
+}
+
+type bakeMetadata map[string]buildStatus
+
+type buildStatus struct {
+	Digest string `json:"containerimage.digest"`
+}
+
+func (s *composeService) doBuildBuildkit(ctx context.Context, options build.Options, ch chan *client.SolveStatus) (string, error) {
+	cfg := bakeConfig{
+		Groups:  map[string]bakeGroup{},
+		Targets: map[string]bakeTarget{},
+	}
+	var group bakeGroup
+	cfg.Targets["build"] = bakeTarget{
+		Context:    options.Inputs.ContextPath,
+		Dockerfile: options.Inputs.DockerfilePath,
+		Args:       options.BuildArgs,
+		Labels:     options.Labels,
+		Tags:       options.Tags,
+		// CacheFrom:  TODO
+		// CacheTo:    TODO
+		// Platforms:  TODO
+		Target: options.Target,
+		// Secrets:    TODO
+		// SSH:        TODO
+		Pull:    options.Pull,
+		NoCache: options.NoCache,
 	}
+	group.Targets = append(group.Targets, "build")
+	cfg.Groups["default"] = group
 
-	for _, img := range response {
-		if img == nil || len(img.ExporterResponse) == 0 {
-			continue
-		}
-		digest, ok := img.ExporterResponse["containerimage.digest"]
-		if !ok {
-			continue
-		}
-		return digest, nil
+	b, err := json.Marshal(cfg)
+	if err != nil {
+		return "", err
 	}
 
-	return "", fmt.Errorf("buildkit response is missing expected result for %s", service)
-}
+	metadata, err := os.CreateTemp(os.TempDir(), "compose")
+	if err != nil {
+		return "", err
+	}
 
-func (s composeService) dryRunBuildResponse(ctx context.Context, name string, options build.Options) map[string]*client.SolveResponse {
-	w := progress.ContextWriter(ctx)
-	buildResponse := map[string]*client.SolveResponse{}
-	dryRunUUID := fmt.Sprintf("dryRun-%x", sha1.Sum([]byte(name)))
-	w.Event(progress.Event{
-		ID:     " ",
-		Status: progress.Done,
-		Text:   fmt.Sprintf("build service %s", name),
-	})
-	w.Event(progress.Event{
-		ID:     "==>",
-		Status: progress.Done,
-		Text:   fmt.Sprintf("==> writing image %s", dryRunUUID),
-	})
-	w.Event(progress.Event{
-		ID:     "==> ==>",
-		Status: progress.Done,
-		Text:   fmt.Sprintf(`naming to %s`, options.Tags[0]),
+	cmd := exec.CommandContext(ctx, "docker", "buildx", "bake", "--file", "-", "--progress", "rawjson", "--metadata-file", metadata.Name())
```

Comment:
I wonder if we should call buildx in standalone mode to reduce overhead implied by `docker` command (and in particular docker proxy in DD).

Reply:
This also has implications wrt signal handling, CLI plugin management, etc. In this case we'd have `CLI binary` -> `Compose plugin binary` -> `CLI binary` -> `Buildx plugin binary` AND between the CLI and Compose and then between the 2nd CLI process and Buildx the CLI would be setting up a socket for plugin communication, etc. Need to think about what the tradeoffs are there/whether we should and what we can do on the CLI side to make this as painless as possible.

Reply:
yes this I already identified, marked this as a Draft to avoid any confusion: there's a bunch of things missing here :)